cpp/src/strings/replace/multi.cu

```diff
@@ -0,0 +1,488 @@
+   * @param d_targets_offsets Offsets identify which target positions go with the current string
+   * @return Number of substrings resulting from the replace operations on this row
+   */
+  __device__ size_type count_strings(size_type idx,
+                                     target_pair const* d_positions,
+                                     size_type const* d_targets_offsets) const
+  {
+    if (!is_valid(idx)) { return 0; }
+
+    auto const d_str             = get_string(idx);
+    auto const d_str_end         = d_str.data() + d_str.size_bytes();
+    auto const base_ptr          = get_base_ptr();
+    auto const targets_positions = cudf::device_span<target_pair const>(
+      d_positions + d_targets_offsets[idx], d_targets_offsets[idx + 1] - d_targets_offsets[idx]);
+
+    size_type count = 1;  // always at least one string
+    auto str_ptr    = d_str.data();
+    for (auto d_pair : targets_positions) {
+      auto const d_pos   = d_pair.first;
+      auto const d_tgt   = d_targets[d_pair.second];
+      auto const tgt_ptr = base_ptr + d_pos;
+      if (str_ptr <= tgt_ptr && tgt_ptr < d_str_end) {
+        auto const keep_size = static_cast<size_type>(thrust::distance(str_ptr, tgt_ptr));
+        if (keep_size > 0) { count++; }  // don't bother counting empty strings
+
+        auto const d_repl = get_replacement_string(d_pair.second);
+        if (!d_repl.empty()) { count++; }
+
+        str_ptr += keep_size + d_tgt.size_bytes();
+      }
+    }
+
+    return count;
+  }
+
+  /**
+   * @brief Retrieve the strings for each row
+   *
+   * This will return string segments as string_index_pair objects for
+   * parts of the string that are not replaced interlaced with the
+   * appropriate replacement string where replacement targets are found.
+   *
+   * This function is called only once to produce both the string_index_pair objects
+   * and the output row size in bytes.
+   *
+   * @param idx Index of the row in d_strings
+   * @param d_offsets Offsets to identify where to store the results of the replace for this string
+   * @param d_positions The target positions found in the chars column
+   * @param d_targets_offsets The offsets to identify which target positions go with this string
+   * @param d_all_strings The output of all the produced string segments
+   * @return The size in bytes of the output string for this row
+   */
+  __device__ size_type get_strings(size_type idx,
+                                   size_type const* d_offsets,
+                                   target_pair const* d_positions,
+                                   size_type const* d_targets_offsets,
+                                   string_index_pair* d_all_strings) const
+  {
+    if (!is_valid(idx)) { return 0; }
+
+    auto const d_output       = d_all_strings + d_offsets[idx];
+    auto const d_output_count = d_offsets[idx + 1] - d_offsets[idx];
+
+    auto const d_str = get_string(idx);
+    if (d_output_count == 1) {
```

Comment:
This is pretty magical. Did I miss the explanation for it? In my past life I would suggest a `#define` for this magic number to give it a good name.

Reply:
There is nothing magical. Unless the input string is null, there will always be at least one output string for any replace operation.

Reply:
Why do we have this if though if the output size is 1? I don't understand why 1 output size of 1 means to grab the entire string.